Remove common.PORT usage from `parallel` tests

adityaanandmc opened the pull request:
Refs: #12376

Tests Updated: 
- test/parallel/test-net-better-error-messages-port-hostname.js
- test/parallel/test-net-socket-destroy-twice.js
- test/parallel/test-net-options-lookup.js 
- test/parallel/test-net-connect-immediate-destroy.js
- test/parallel/test-net-localerror.js
- test/parallel/test-net-connect-handle-econnrefused.js

Tests ported from test/parallel to test/sequential:
- test/sequential/test-net-listen-shared-ports.js
- test/sequential/test-net-better-error-messages-port.js

Analyzed and continuing to use common.PORT
 - test/parallel/test-net-connect-immediate-finish.js 

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
test

adityaanandmc commented:
First-time contributor. (Noob level)
Anything I'm missing out on?

vsemozhetbyt commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/7455/

thefourtheye commented:
Good job @adityaandmc :clap:

thefourtheye commented:
@adityaanandmc Changes are all good. If you could make sure if the commits follow these [guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-guidelines) it would be great?

adityaanandmc commented:
@thefourtheye Do the changes look alright ?

adityaanandmc commented:
@thefourtheye @cjihrig  
Are we allowed to do this? Create a server and kill it simply to get a free port.
Is there a way to just get an unused port ?

cjihrig commented:
> Are we allowed to do this? Create a server and kill it simply to get a free port.

That would probably introduce a race condition. Once you close the server to release the port, something else can claim that port.

adityaanandmc commented:
@cjihrig What would be the ideal way to get this done? I was aware of the race condition and hence a little stuck. My commit ( code below ) will fail if the race condition is met.
How do I secure a free port and listen to it ensuring that a server hasn't been created in the gap?

```
const net = require('net');
const assert = require('assert');

//const c = net.createConnection(common.PORT);
const server = net.createServer();
server.listen(0);
const port = server.address().port;
server.close();

const c = net.createConnection(port);

c.on('connect', common.mustNotCall());

c.on('error', common.mustCall(function(e) {
    assert.strictEqual(e.code, 'ECONNREFUSED');
    //assert.strictEqual(e.port, common.PORT);
    assert.strictEqual(e.port, port);
    assert.strictEqual(e.address, '127.0.0.1');
}));
```

thefourtheye commented:
@adityaanandmc wrote

> How do I secure a free port and listen to it ensuring that a server hasn't been created in the gap?

Although the possibility of this happening is less, it is still possible and we cannot avoid it.

santigimeno commented:
@adityaanandmc what about doing something like this:

```js
...
const server = net.createServer();
server.listen(0);
const port = server.address().port;
const c = net.createConnection(port);
server.close();
...
```

adityaanandmc commented:
@santigimeno  Now i feel like a idiot :P Thanks for that!!

santigimeno commented:
@adityaanandmc could you squash the commits into one? Thanks!

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/7523/

adityaanandmc commented:
@santigimeno will squash them!! 
Looks like there are a few failures [test-net-connect-immediate-finish](https://github.com/nodejs/node/pull/12473/commits/0c2dd7e6c3fe2cfc3460fa954102241216479a12)

I can't seem to understand the failures. I thought about doing this, but it didn't seem like the right thing to do.

```
const server = net.createServer();
server.listen(0);
const port = server.address().port;
server.close();

const client = net.connect({host: '***', port: port});
```

santigimeno commented:
@adityaanandmc looking at the changes again it seems that we're trying to `connect()` to the `0` port. I don't think that makes sense: the port `0` makes sense when calling `bind()` or `listen()` not `connect()`. What I think it could be done is create a helper function in `common.js` (connectToRandomPort for example) to reserve a port and connect, similar to the code you were proposing.

adityaanandmc commented:
@santigimeno I did that and it passed locally but at times it fails with the following error!
It started only after i had touched this file [test-net-connect-immediate-finish](https://github.com/nodejs/node/pull/12473/commits/582ef6241e5fb36d63f142e01cb8b8c099e3ea4f). 
Not able to figure out why it's happening. 

```
Path: parallel/test-domain-no-error-handler-abort-on-uncaught-0
Command: out/Release/node /Users/adianand/Labs/node/test/parallel/test-domain-no-error-handler-abort-on-uncaught-0.js
--- TIMEOUT ---
=== release test-domain-no-error-handler-abort-on-uncaught-1 ===
Path: parallel/test-domain-no-error-handler-abort-on-uncaught-1
Command: out/Release/node /Users/adianand/Labs/node/test/parallel/test-domain-no-error-handler-abort-on-uncaught-1.js
--- TIMEOUT ---
=== release test-domain-no-error-handler-abort-on-uncaught-2 ===
Path: parallel/test-domain-no-error-handler-abort-on-uncaught-2
Command: out/Release/node /Users/adianand/Labs/node/test/parallel/test-domain-no-error-handler-abort-on-uncaught-2.js
--- TIMEOUT ---
=== release test-domain-no-error-handler-abort-on-uncaught-3 ===
Path: parallel/test-domain-no-error-handler-abort-on-uncaught-3
Command: out/Release/node /Users/adianand/Labs/node/test/parallel/test-domain-no-error-handler-abort-on-uncaught-3.js
--- TIMEOUT ---
=== release test-domain-no-error-handler-abort-on-uncaught-4 ===
Path: parallel/test-domain-no-error-handler-abort-on-uncaught-4
Command: out/Release/node /Users/adianand/Labs/node/test/parallel/test-domain-no-error-handler-abort-on-uncaught-4.js
--- TIMEOUT ---
```

Sorry for flooding you with doubts. Just a lil confused for a first-time contributor on a project of this scale.

santigimeno commented:
@adityaanandmc no worries.

I've been looking into this again. The tests: `test-net-better-error-messages-port-hostname`, `test-net-connect-immediate-destroy`, `test-net-options-lookup` and `test-net-connect-immediate-finish` doesn't really get to try to bind the socket to a specific port, so using port `0`, `common.PORT` or any other port should be ok. So I *think* every test should be ok as it is except for `test-net-connect-immediate-finish` that should be fine without the `server` code and just trying to connect to some port. I hope that makes the tests pass again and sorry for the confusion.

adityaanandmc commented:
@Trott Would it make sense to stick to common.PORT for test-net-connect-immediate-finish ?

Trott commented:
> @Trott Would it make sense to stick to common.PORT for test-net-connect-immediate-finish ?
If we are hell bent on removing common.PORT, I could just set the port to 12346 or another random port to avoid the TIMEOUT error.

Might need a case-by-case evaluation, but in situations where you might be tempted to hardcode the port, it might generally be better to move the test from `parallel` to `sequential` and continue using `common.PORT`. 

For that matter, it may make sense for `common.PORT` to stick around in a few `parallel` tests where the port number is a placeholder and nothing actually listens or tries to connect to that port. (Like, if the hostname is supposed to be garbled and cause things to fail.) Even if we want a lint rule in that case, we can just disable the lint rule for that one line with a comment.

Not sure I'm really answering your question here. But if you think it might make sense to keep `common.PORT`, then maybe we should keep it here.

adityaanandmc commented:
In that case, leaving as it is!! Thanks

santigimeno commented:
@adityaanandmc I wouldn't move `test-net-connect-immediate-finish ` to sequential. Using `common.PORT` does not cause a concurrency problem as the `bind()` call is never performed, or as @Trott said:

> For that matter, it may make sense for common.PORT to stick around in a few parallel tests where the port number is a placeholder and nothing actually listens or tries to connect to that port. (Like, if the hostname is supposed to be garbled and cause things to fail.) Even if we want a lint rule to flag common.PORt in parallel tests, we can just disable the lint rule for that one line with a comment.

santigimeno commented on test/parallel/test-net-listen-shared-ports.js:
```diff
@@ -54,10 +54,11 @@ if (cluster.isMaster) {
   }, function() {
-    server2.listen({port: common.PORT + 1, exclusive: true}, function() {
+    const port2 = server1.address().port;
+    server2.listen({port: port2, exclusive: true}, function() {
```
`server2` is listening in the same port as `server1`. It doesn't look right

adityaanandmc replied:
Ah my bad

adityaanandmc replied:
@santigimeno I've made the change , could you verify it ?

Trott commented on test/parallel/test-net-better-error-messages-port-hostname.js:
```diff
@@ -3,12 +3,12 @@ const common = require('../common');
 const assert = require('assert');
 
-const c = net.createConnection(common.PORT, '***');
+const c = net.createConnection(0, '***');
```
`createConnection()` can not take port 0 the way `listen()` and `bind()` can. Sending port 0 to createConnection() is an error.

Trott replied:
Specifically, sending to a valid port that's not listening will end up getting `ECONNREFUSED` but sending port 0 will end up getting `EADDRNOTAVAIL`.

adityaanandmc replied:
Makes sense! Will fix it

Trott commented on test/parallel/test-net-better-error-messages-port-hostname.js:
```diff
@@ -3,12 +3,18 @@ const common = require('../common');
-const c = net.createConnection(common.PORT, '***');
+const s = net.createServer();
+s.listen(0);
+const port = s.address().port;
```
Since the `createConnection()` never gets a proper hostname, the actual connection never happens, so we don't need to carefully choose an available port, I don't think. I think you can just hardcode any valid port (e.g., `80`) and that should be fine. Maybe include a comment explaining that the port needs to be valid but is never in fact used because a connection never gets created because the hostname is invalid.

thefourtheye replied:
I think we are overthinking here. As the port doesn't matter for this test, we can simply do the following and be done with it.

```diff
diff --git a/test/parallel/test-net-better-error-messages-port-hostname.js b/test/parallel/test-net-better-error-messages-port-hostname.js
index 14f84eb15b..896fe29b38 100644
--- a/test/parallel/test-net-better-error-messages-port-hostname.js
+++ b/test/parallel/test-net-better-error-messages-port-hostname.js
@@ -3,12 +3,12 @@ const common = require('../common');
 const net = require('net');
 const assert = require('assert');

-const c = net.createConnection(common.PORT, '***');
+const c = net.createConnection(0, '***');

 c.on('connect', common.mustNotCall());

 c.on('error', common.mustCall(function(e) {
   assert.strictEqual(e.code, 'ENOTFOUND');
-  assert.strictEqual(e.port, common.PORT);
+  assert.strictEqual(e.port, 0);
   assert.strictEqual(e.hostname, '***');
 }));
```

Trott replied:
@thefourtheye Might be overthinking it indeed, but I prefer a valid port because IMO selecting port `0` for `createConnection()` is undefined behavior, so I'd rather not codify the expectations in tests. It's also kind of misleading because someone reading the code might infer that it selects an arbitrary port but it doesn't. So I'd rather see a valid port, and any valid port will do, so `80` or `443` or `8080` or whatever would all be fine by me. 

All that said, you're correct that what you suggest above will work and if you feel that `0` is the way to go and @adityaanandmc agrees, I won't object. I have opinions, but not strong ones.

adityaanandmc replied:
@Trott If that's the case, I would prefer using common.PORT instead rather.
I figured I'd stick with 0 cause like you said `the actual connection never happens` but I'm okay going either way.

thefourtheye replied:
@Trott We are asserting that the port is zero, right? That makes sure that no arbitrary port is assigned. We can probably leave a comment explaining it.

Trott replied:
I'm fine with `common.PORT` or a hard-coded number whether it be `8080`,  `0`, or whatever. Like I said, I have opinions, but not strong ones.

adityaanandmc commented:
@cjihrig @Trott @santigimeno 
You guys probably get this a lot. (I'd like to contribute more than just fixing a few tests )
Figured going through the internals of node would give me some insight.
Any advice on where to start would be much appreciated ?

santigimeno commented on test/parallel/test-net-listen-shared-ports.js:
```diff
@@ -42,22 +42,23 @@ if (cluster.isMaster) {
-    server2.listen({port: common.PORT + 1, exclusive: true}, function() {
+  }, () => {
+    const port2 = server1.address().port + 1;
+    server2.listen({port: port2, exclusive: true}, () => {
```
I think there's still a problem here. There's not guarantee that `port2` is free when you call listen. See https://github.com/nodejs/node/pull/12452#discussion_r111692779 for a similar issue

santigimeno commented:
> You guys probably get this a lot. (I'd like to contribute more than just fixing a few tests )
Figured going through the internals of node would give me some insight.
Any advice on where to start would be much appreciated ?

I don't have a good answer myself. I guess looking for issues labeled as [good first contribution](https://github.com/nodejs/node/labels/good%20first%20contribution ) can be a good starting point.

Trott commented:
>  (I'd like to contribute more than just fixing a few tests )
Figured going through the internals of node would give me some insight.
Any advice on where to start would be much appreciated ?

Take a look at http://nodetodo.org/next-steps/

adityaanandmc commented:
@santigimeno [test/parallel/test-net-listen-shared-ports.js](https://github.com/nodejs/node/pull/12473/files#diff-3a4169e6b1957d57fbeaaf6b329c1c30) has been updated. Thanks for the reference. Didn't want to port it over to sequence. Is it alright to change the file to this extent ? 

The ports are determined within the master process and the workers then use these ports.

santigimeno commented:
@adityaanandmc I still think there's a race in the implementation. By the time the ports are used in the workers they could have been taken by another process. Probably the best solution is going to be moving the test to sequential.

adityaanandmc commented:
@santigimeno I was hell bent on trying to get it to stay in parallel. The last option was to pass the server instead of the just the ports after closing the server to avoid anyone else grabbing the port but that would make the test unnecessarily complicated. I'd reckon that's a bad idea !

So stick to sequential ?

santigimeno commented:
@adityaanandmc I *think* a solution might be that the first worker makes the servers listen on random ports, then pass the ports back to master, then the master passes the ports to the second worker that uses them. That could be too complicated as well, though, so probably moving it to sequential is just fine.

adityaanandmc commented:
@santigimeno Ported it over to sequential

santigimeno commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/7882/

thefourtheye commented:
CI Run: https://ci.nodejs.org/job/node-test-pull-request/7889/

adityaanandmc commented:
@santigimeno  had an test failure in `parallel/test-net-connect-immediate-destroy` cause I forgot to change `net.connect(0, ...)` to a valid port. Could you have a [look at it](https://github.com/nodejs/node/pull/12473/files#diff-b8be9e852bd3b457a0d160ad68268197) if it's alright and then trigger a build if so ?

santigimeno commented on test/parallel/test-net-connect-immediate-destroy.js:
```diff
@@ -2,7 +2,11 @@
+  const socket = net.connect(port, common.localhostIPv4, common.mustNotCall());
+  socket.on('error', common.mustNotCall());
+  socket.destroy();
+});
```
I would close the server **after** calling `net.connect()` just to be sure no other process has reclaimed the free port.
```js
'use strict';
const common = require('../common');
const net = require('net');

const s = net.createServer();
s.listen(0);
const port = s.address().port;
const socket = net.connect(port, common.localhostIPv4, common.mustNotCall());
socket.on('error', common.mustNotCall());
s.close();
socket.destroy();
```

adityaanandmc replied:
@santigimeno  Is that right though ? 
The original test tried to connect to `common.PORT` without creating a server thus ensuring that the connectListener `common.mustNotCall()` would not be invoked.
If i were to ensure that a server was listening on the port , that would invoke `common.mustNotCall()` => which is not what this test expects , right? 
(I'm trying to wrap my head around it, apologies if the questions seem silly)

santigimeno replied:
I don't think it's a silly question at all. I _think_ that closing the socket right after calling connect should work because the connect operation is asynchronous and we're closing the socket synchronously (I haven't tested on every platform though just locally on `OS X`, I started a CI job with that changes to see if it works: https://ci.nodejs.org/job/node-test-commit/9693/https://ci.nodejs.org/job/node-test-commit/9693/). OTOH I would argue that creating a server just to reserve a port might make the code too convoluted and moving it to sequential could be a better option.

adityaanandmc replied:
Well, that it seems to be working. I'll change the code accordingly!

Creating a server just to secure a port did seem unnecessary but would it be alright if still kept it under parallel?

santigimeno replied:
> Creating a server just to secure a port did seem unnecessary but would it be alright if still kept it under parallel?

I think it's the only way if we want to keep it under parallel. BTW, I've just noticed there are other tests in the PR with the same issue: attempting the connection after closing the server. Could you update those too? Then I'll launch a CI job. Thanks"

adityaanandmc replied:
Yup! Updated.

santigimeno commented:
Let's see if it comes out green: https://ci.nodejs.org/job/node-test-pull-request/7912/

adityaanandmc commented:
I am not having a good day 😛 

@santigimeno So looking into the [logs](https://ci.nodejs.org/job/node-test-binary-windows/RUN_SUBSET=3,VS_VERSION=vs2015-x86,label=win2008r2/8289/consoleText), this is what i see : 

```
c:\workspace\node-test-binary-windows\RUN_SUBSET\3\VS_VERSION\vs2015-x86\label\win2008r2>TASKKILL /F /IM node.exe /T   || TRUE
ERROR: The process "node.exe" not found.

c:\workspace\node-test-binary-windows\RUN_SUBSET\3\VS_VERSION\vs2015-x86\label\win2008r2>TASKKILL /F /IM cctest.exe /T   || TRUE
ERROR: The process "cctest.exe" not found.

c:\workspace\node-test-binary-windows\RUN_SUBSET\3\VS_VERSION\vs2015-x86\label\win2008r2>TASKKILL /F /IM run-tests.exe /T   || TRUE
ERROR: The process "run-tests.exe" not found.
```
Under what scernio does this happen ? I'm not sure what i messed up

addaleax commented:
I think the actual error in the output you linked to is this:

```
not ok 357 sequential/test-benchmark-child-process
  ---
  duration_ms: 60.190
  severity: fail
  stack: |-
    timeout
```

That is unrelated, and was tracked at https://github.com/nodejs/node/issues/12817 (but *should* be fixed now).

Either way, it looks like this PR has a merge conflict now so it might need to rebase – is that something you can do?

santigimeno commented:
@adityaanandmc, what @addaleax said.

BTW, could you address this [nit](https://github.com/nodejs/node/pull/12473#pullrequestreview-36454466). Thanks!

adityaanandmc commented:
Sure thing !! Will do that in a bit  though. ( heading out to watch Guardians of the Galaxy ) :D

adityaanandmc commented:
@addaleax Rebased!!

gibfahn commented:
CI: https://ci.nodejs.org/job/node-test-commit/9717/

santigimeno commented:
@adityaanandmc sorry to bring this up again, but it would be great if you could address [this](https://github.com/nodejs/node/pull/12473#pullrequestreview-36454466). Anyway, the PR as it is LGTM. Thanks!

adityaanandmc commented:
@santigimeno That's alright. I did not see that comment until now. Sorry about that!
Added comments ! Do they look alright ? 

Updated: 
- [test-net-better-error-messages-port-hostname](https://github.com/nodejs/node/pull/12473/files#diff-15a47705a207bddd6a5ccb3e02319744R6)
- [test-net-options-lookup.js](https://github.com/nodejs/node/pull/12473/files#diff-c4567f525f6ac0b0e35f3df656f62240R10)

Thanks @santigimeno @thefourtheye @Trott @cjihrig !!

santigimeno commented:
> Added comments ! Do they look alright ?

Yes, LGTM. I think another in `test-net-localerror.js` could be useful. Thanks!

adityaanandmc commented:
@santigimeno Done!

thefourtheye commented on test/parallel/test-net-better-error-messages-port.js:
```diff
@@ -3,12 +3,14 @@ const common = require('../common');
+  assert.strictEqual(e.port, port);
   assert.strictEqual(e.address, '127.0.0.1');
 }));
+server.close();
```
Okay, `server.close` is actually asynchronous. Perhaps we should run this test in the `close` event of `server`.

santigimeno replied:
> Okay, server.close is actually asynchronous

You're right, though the `close()` is executed syncronously at least on Unixes, but that's an implementation detail.

> Perhaps we should run this test in the close event of server.

I think that's not going to work either, because there's going to be a window when the port is not assigned and could be grabbed by some other process/test. What do you think @thefourtheye ?

So probably the only way to make this work without race conditions is moving it to sequential.

thefourtheye replied:
@santigimeno `close()` is actually executed synchronously, but by the time the connection attempt is made (which happens asynchronously), the port is still up for grab, right? Perhaps I am overthinking 🤔 

I am afraid sequential also will not solve this. `common.PORT` could still have been assigned to other processes.

Perhaps we can live with the current change...

santigimeno replied:
> @santigimeno close() is actually executed synchronously, but by the time the connection attempt is made (which happens asynchronously), the port is still up for grab, right? Perhaps I am overthinking 🤔

Yes, I thought of that too. And I think I'm also overthinking 😆 .

> I am afraid sequential also will not solve this. common.PORT could still have been assigned to other processes.

Right, but not to other tests in the same run though.

> Perhaps we can live with the current change...

Yeah, probably they'll never fail.

adityaanandmc replied:
Since we've reached a moo point. Would it be alright if I followed @thefourtheye approach and change the code accordingly ?

thefourtheye replied:
@adityaanandmc No, move this to sequential, without the changes.

adityaanandmc commented:
@thefourtheye Done. CI Build trigger necessary ?

thefourtheye commented:
CI Run: https://ci.nodejs.org/job/node-test-pull-request/7965/

jasnell commented:
Landed in 94eed0f
